core/Services/Calendar/Store.vala

```diff
@@ -26,282 +26,630 @@ public class Calendar.Store : Object {
 
-    /* Notifies when events are added, updated, or removed */
-    public signal void events_added (E.Source source, Gee.Collection<ECal.Component> events);
-    public signal void events_updated (E.Source source, Gee.Collection<ECal.Component> events);
-    public signal void events_removed (E.Source source, Gee.Collection<ECal.Component> events);
+    /* Notifies when components are added, modified, or removed */
+    public signal void components_added (Gee.Collection<ECal.Component> components, E.Source source, Gee.Collection<ECal.ClientView> views);
+    public signal void components_modified (Gee.Collection<ECal.Component> components, E.Source source, Gee.Collection<ECal.ClientView> views);
+    public signal void components_removed (Gee.Collection<ECal.Component> components, E.Source source, Gee.Collection<ECal.ClientView> views);
 
-    public signal void connecting (E.Source source, Cancellable cancellable);
-    public signal void connected (E.Source source);
-    public signal void error_received (string error);
+    /* Notifies when sources are added, changed, or removed */
+    public signal void source_connecting (E.Source source, GLib.Cancellable cancellable);
+    public signal void source_added (E.Source source);
+    public signal void source_changed (E.Source source);
+    public signal void source_removed (E.Source source);
+
+    public signal void error_received (GLib.Error e);
 
     /* The month_start, num_weeks, or week_starts_on have been changed */
     public signal void parameters_changed ();
 
-    HashTable<string, ECal.Client> source_client;
-    HashTable<string, ECal.ClientView> source_view;
-    HashTable<E.Source, Gee.TreeMultiMap<string, ECal.Component>> source_events;
+    public ECal.ClientSourceType source_type { get; construct; }
+    private HashTable<string, ECal.Client> source_client;
+    private HashTable<string, Gee.ArrayList<ECal.ClientView>> source_views;
+
+    private HashTable<ECal.ClientView, Gee.Collection<ECal.Component>> components_add_transaction;
+    private Gee.Collection<E.Source> sources_add_transaction;
+    private Gee.Collection<E.Source> sources_connect_transaction;
+
+    internal HashTable<string, Gee.TreeMultiMap<string, ECal.Component>> source_components;
 
-    public GLib.Queue<E.Source> calendar_trash;
+    private GLib.Queue<E.Source> sources_trash;
     private E.CredentialsPrompter credentials_prompter;
 
-    private static Calendar.Store? store = null;
     private static GLib.Settings state_settings;
 
-    public static Calendar.Store get_default () {
-        if (store == null)
-            store = new Calendar.Store ();
-        return store;
+    private Store (ECal.ClientSourceType source_type) {
+        Object (source_type: source_type);
+    }
+
+    private static Calendar.Store? event_store = null;
+    private static Calendar.Store? task_store = null;
+
+    public static Calendar.Store get_event_store () {
```

Comment:
I'm not sure this is more clear. The convention of `get_default` is used across many apps and libraries and makes it more immediately clear that this is a singleton

Reply:
I think the issue here is `get_event_store` vs `get_task_store`. Since there can be both, there has to be some way of distinguishing this.

Reply:
@mcclurgm is right; actually we are dealing with two singletons here: one for tasks and one for events. However, we can change the API to something like:

- `get_default (ECal.ClientSourceType store_type)`

or:

- `get_default_event_store ()`
- `get_default_task_store ()`

if something like this is preferred. What do you think @danrabbit?

Reply:
Oh interesting. Sorry I'm just reading through this. Is the task store used at all in Calendar?

Reply:
@danrabbit thats fine. The task_store is not used in the Calendar app. An example of both stores are in use in wingpanel-indicator-datetime - even though the Calendar.Store there is a bit older.

Reply:
FWIW: we can also use an abstract super class "Calendar.Store" with two implementations: "Calendar.TaskStore", "Calendar.EventStore" which then can expose a corresponding "get_default ()".

Reply:
Just thinking out loud, if we wanted to go that direction I can see a compelling argument for making the generic Store an abstract class and embracing polymorphism. Other functions like `view_icalcomponents_added` have branches for the different types, and it would be more readable to make them implementation details of concrete subclasses instead of adding a branching if statement with another level of indentation.

Reply:
@mcclurgm yeah, the code base might really benefit from this separation. Also I guess it would make it easier to differentiate between more Tasks or Calendar specific code. So I guess I really like the idea of having such a thing like:

```
public abstract class Calendar.Store {
  ...
}

public class Calendar.EventStore: Calendar.Store {
  public static Calendar.EventStore get_default () {
    ...
  }
  ...
}

public class Calendar.TaskStore: Calendar.Store {
  public static Calendar.TaskStore get_default () {
    ...
  }
  ...
}
```

Reply:
That's exactly what I was thinking. If we have a nice OO language, might as well use it :nerd_face: 

I could try to take that on or help you along once this is merged. I assume it would be better to keep the design as is for now and do more implementation detail changes in a separate review.